homeplug_exporter.go

```diff
@@ -291,29 +290,17 @@ ChanLoop:
 }
 
 func read_homeplug(conn *packet.Conn, ch chan<- HomeplugNetworkInfo) {
-	// The HomePlug AV specification limits the size of management messages to 1518 bytes.
-	b := make([]byte, 1518)
-
 	for {
-		conn.SetReadDeadline(time.Now().Add(time.Second))
-		n, addr, err := conn.ReadFrom(b)
+		addr, f, err := readFrame(conn)
```

Comment:
We could probably drop `addr`, as the same value will be present in `f.Source`

Reply:
I'm 99% inclined to agree. 1% of me says there may be corner cases where this might not necessarily be the case, but I shall silence those doubts for now. For sure, you can pass NULL `src_addr` and `addrlen` to `recvfrom(2)` (which is the syscall that mdlayer/packet ultimately wraps), in which case they will not be populated - so that appears to be a legitimate use case. The main reason AFAICT for the source addr to be returned from [Conn.ReadFrom](https://pkg.go.dev/github.com/mdlayher/packet#Conn.ReadFrom) is that it needs to implement that method to satisfy the [net.PacketConn](https://pkg.go.dev/net#PacketConn) interface.

Reply:
If `src_addr` were null (but I don' t know if the mdlayer code allows it) that should affect the returned address, but the contents of the ethernet frame should be the canonical source for that information, unless there is some bad bug in the kernel, no?

Reply:
I was merely referring to the libc `recvfrom` function, which is obviously just a wrapper for the `recvfrom` syscall.

> If  src_addr  is  not  NULL,  and  the  underlying  protocol  provides the source address of the message, that source address is placed in the buffer pointed to by src_addr. In this case, addrlen is a value-result argument. Before the call, it should be initialized to the size of the buffer associated with  src_addr. Upon return, addrlen is updated to contain the actual size of the source address. The returned address is truncated if the buffer provided is too small; in this case, addrlen will return a value greater than was supplied to the call.
>
> If the caller is not interested in the source address, src_addr and addrlen should be specified as NULL.

In Go, [packet.Conn.ReadFrom](https://pkg.go.dev/github.com/mdlayher/packet#Conn.ReadFrom) is a bit of a long call chain, deferring to [socket.Conn.Recvfrom](https://pkg.go.dev/github.com/mdlayher/socket#Conn.Recvfrom), which calls [unix.Recvfrom](https://pkg.go.dev/golang.org/x/sys/unix#Recvfrom), which wraps the `recvfrom` syscall, and indeed should return a source addr:
```go
func Recvfrom(fd int, p []byte, flags int) (n int, from Sockaddr, err error) {
	var rsa RawSockaddrAny
	var len _Socklen = SizeofSockaddrAny
	if n, err = recvfrom(fd, p, flags, &rsa, &len); err != nil {
		return
	}
	if rsa.Addr.Family != AF_UNSPEC {
		from, err = anyToSockaddr(fd, &rsa)
	}
	return
}

...

func recvfrom(fd int, p []byte, flags int, from *RawSockaddrAny, fromlen *_Socklen) (n int, err error) {
	var _p0 unsafe.Pointer
	if len(p) > 0 {
		_p0 = unsafe.Pointer(&p[0])
	} else {
		_p0 = unsafe.Pointer(&_zero)
	}
	r0, _, e1 := Syscall6(SYS_RECVFROM, uintptr(fd), uintptr(_p0), uintptr(len(p)), uintptr(flags), uintptr(unsafe.Pointer(from)), uintptr(unsafe.Pointer(fromlen)))
	n = int(r0)
	if e1 != 0 {
		err = errnoErr(e1)
	}
	return
}
```
The source MAC address in the frame coming in off the wire _should_ be the canonical source for that information. The only way I could see the `src_addr` populated by `recvfrom` being different would be if something funky happened to the sk_buff in kernel space after the frame was copied from the RX ring of the NIC driver.